productivity-suite/piercing-library/src/piercing-gateway.ts

```diff
@@ -135,30 +144,33 @@ export class PiercingGateway<Env> {
       ?.includes("text/html");
 
     if (requestIsForHtml) {
-      const baseUrl = this.config.getBaseAppUrl(env).replace(/\/$/, "");
+      const baseUrl = this.config.getLegacyAppBaseUrl(env).replace(/\/$/, "");
       const indexBodyResponse = this.fetchBaseIndexHtml(
         new Request(baseUrl, request)
       ).then((response) => response.text());
 
+      const piercingEnabled =
+        !this.config.shouldPiercingBeEnabled ||
+        (await this.config.shouldPiercingBeEnabled(request, env, ctx));
+
       const fragmentStreamOrNullPromises: Promise<ReadableStream | null>[] =
-        Array.from(this.fragmentConfigs.values()).map((fragmentConfig) => {
-          const shouldBeIncluded = fragmentConfig.shouldBeIncluded(
-            request,
-            env,
-            ctx
-          );
-
-          const shouldBeIncludedPromise =
-            shouldBeIncluded instanceof Promise
-              ? shouldBeIncluded
-              : new Promise<boolean>((resolve) => resolve(shouldBeIncluded));
-
-          return shouldBeIncludedPromise.then((shouldBeIncluded) =>
-            shouldBeIncluded
-              ? this.fetchSSRedFragment(env, fragmentConfig, request)
-              : null
-          );
-        });
+        !piercingEnabled
+          ? []
+          : Array.from(this.fragmentConfigs.values()).map((fragmentConfig) => {
+              const shouldBeIncluded = fragmentConfig.shouldBeIncluded(
+                request,
+                env,
+                ctx
+              );
+
+              return new Promise(async (resolve) =>
+                resolve(await shouldBeIncluded)
+              ).then((shouldBeIncluded) =>
+                shouldBeIncluded
+                  ? this.fetchSSRedFragment(env, fragmentConfig, request)
+                  : null
+              );
```

Comment:
How about:

```ts
        : Array.from(this.fragmentConfigs.values()).map(
              async (fragmentConfig) => {
                const shouldBeIncluded = await fragmentConfig.shouldBeIncluded(
                  request,
                  env,
                  ctx
                );

                return shouldBeIncluded
                  ? this.fetchSSRedFragment(env, fragmentConfig, request)
                  : null;
              }
            );

```

Reply:
ah yeah, I was thinking of something like this but I was wondering if it could mess up the order of the promises (I played a bit with it and it doesn't seem to)

anyways the order wouldn't be that important either way :+1: